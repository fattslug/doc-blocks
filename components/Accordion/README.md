# @doc-blocks/accordion

Welcome to my `Accordion` component.

## Installation

```sh
npm i @doc-blocks/accordion
# or with yarn
yarn add @doc-blocks/accordion
```

Then to use the component in your code just import it!

```js
import Accordion from "@doc-blocks/accordion";
// and with css-modules
import "@doc-blocks/accordion/dist/main.css";
```

## Usage

The Accordion component handles all click events internally so all you have to worry about is the content you want to display.
Remember to provide an `id` prop to tie the accordion title to its panel!
Feel free to add whatever classes or styles you want to either `Accordion.Title` or `Accordion.Panel` - they will be applied to the respective DOM elements.
To use the `Accordion`, just do this:

```js
const AccordionInterface = () => (
  <Accordion>
    <Accordion.Title id="happy">Colors that make me happy</Accordion.Title>
    <Accordion.Panel id="happy">
      Black, Burgundy, Maroon, Mahogany
    </Accordion.Panel>
    <Accordion.Title id="angry">Colors that make me angry</Accordion.Title>
    <Accordion.Panel id="angry">Pink, Green, White, Orange</Accordion.Panel>
  </Accordion>
);
```

Or with an `Array.map()`:

```js
const AccordionInterface = () => (
  <Accordion>
    {accordionItems.map((accordion) => (
      <React.Fragment key={accordion.id}>
        <Accordion.Title id={accordion.id}>{accordion.title}</Accordion.Title>
        <Accordion.Panel id={accordion.id}>{accordion.content}</Accordion.Panel>
      </React.Fragment>
    ))}
  </Accordion>
);
```

### onChange Handler

When a user clicks on an Accordion title, the `onChange` handler is called. You can make use of this functionality by adding an `onChange` prop to the `<Accordion>` wrapper like so:

```jsx
const TabbedInterface = () => (
  <Accordion onChange={(selectedId) => console.log(selectedId)}>...</Accordion>
);
```

### Active Class

You can apply an `activeClassName` to each `Accordion.Title` and `Accordion.Panel` component. When that accordion item is selected, the specified classes will be applied.

```jsx
const AccordionInterface = () => (
  <Accordion>
    <Accordion.Title id="happy" activeClassName="selected-title-blue">
      Colors that make me happy
    </Accordion.Title>
    <Accordion.Panel id="happy" activeClassName="selected-panel">
      Black, Burgundy, Maroon, Mahogany
    </Accordion.Panel>
    <Accordion.Title id="angry" activeClassName="selected-title-red">
      Colors that make me angry
    </Accordion.Title>
    <Accordion.Panel id="angry" activeClassName="selected-panel">
      Pink, Green, White, Orange
    </Accordion.Panel>
  </Accordion>
);
```
