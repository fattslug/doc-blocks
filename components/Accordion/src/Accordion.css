.accordion {
  position: relative;
  margin: 14px 0;
  width: 100%;
}

.title {
  width: 100%;
  padding: 14px 28px;

  &:hover,
  &:focus {
    background-color: #fafafa;
    cursor: pointer;
  }
}

.panel {
  width: 100%;
  box-sizing: content-box;
  transition: height 0.2s ease-in-out, padding 0.2s ease-in-out;
  height: auto;
  overflow: hidden;

  &.collapsed {
    padding: 0 28px;
  }

  &.expanded {
    padding: 14px 28px;
  }
}

@media (prefers-reduced-motion) {
  .panel {
    transition: none;
  }
}
