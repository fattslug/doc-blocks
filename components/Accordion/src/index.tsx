import React from "react";

import styles from "./Accordion.css";

interface AccordionCompositionProps
  extends React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > {
  /** ID to assign to accordion item */
  id: string;
  /** Class to apply when selected */
  activeClassName?: string;
}

interface AccordionProps {
  /** Children of accordion wrapper */
  children: React.ReactChild[];
}

interface AccordionContextProps {
  /** ID of currently selected panel  */
  selectedId: string | null;
  /** Setter method for setting selected panel */
  setSelectedId: React.Dispatch<React.SetStateAction<string | null>>;
}

// Setting up context with dummy values to satisfy typechecking
const AccordionContext = React.createContext<AccordionContextProps>({
  selectedId: null,
  setSelectedId: () => null,
});

/** Accordion component */
const Accordion = ({ children }: AccordionProps) => {
  const [selectedId, setSelectedId] = React.useState<string | null>(null);

  return (
    <AccordionContext.Provider value={{ selectedId, setSelectedId }}>
      <div className={styles.accordion}>{children}</div>
    </AccordionContext.Provider>
  );
};

/** Renders the accordion title */
const Title = ({
  id,
  children,
  className,
  activeClassName,
}: AccordionCompositionProps) => {
  const { selectedId, setSelectedId } = React.useContext(AccordionContext);

  return (
    <div
      className={`${styles.title} ${className} ${
        selectedId === id ? activeClassName : ""
      }`}
      id={`${id}-title`}
      role="button"
      tabIndex={0}
      onClick={() => setSelectedId((prev) => (prev === id ? null : id))}
      onKeyPress={() => setSelectedId((prev) => (prev === id ? null : id))}
    >
      {children}
    </div>
  );
};

/** Renders the accordion content panel */
const Panel = ({
  id,
  children,
  className,
  activeClassName,
}: AccordionCompositionProps) => {
  const { selectedId } = React.useContext(AccordionContext);
  const panelRef = React.useRef<HTMLDivElement>(null);
  const isSelected = selectedId === id;

  return (
    <div
      ref={panelRef}
      className={`${styles.panel} ${
        isSelected ? styles.expanded : styles.collapsed
      } ${className} ${selectedId === id ? activeClassName : ""}`}
      id={`${id}-panel`}
      aria-hidden={!isSelected}
      style={{
        height: isSelected ? panelRef.current?.scrollHeight : "0px",
      }}
    >
      {children}
    </div>
  );
};

Accordion.Title = Title;
Accordion.Panel = Panel;

export default Accordion;
