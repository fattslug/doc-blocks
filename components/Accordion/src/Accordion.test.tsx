import React from "react";
import { render } from "@testing-library/react";
import Accordion from ".";

describe("Accordion", () => {
  test("It matches the snapshot", () => {
    const { container } = render(<Accordion />);
    expect(container).toMatchSnapshot();
  });
});
