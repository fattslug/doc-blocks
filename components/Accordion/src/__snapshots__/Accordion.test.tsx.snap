// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion It matches the snapshot 1`] = `
<div>
  <div
    class="accordion"
  >
    <div
      class="title title-1-default "
      id="one-title"
      role="button"
      tabindex="0"
    >
      Title 1
    </div>
    <div
      aria-hidden="true"
      class="panel collapsed panel-1-default "
      id="one-panel"
      style="height: 0px;"
    >
      Panel 1
    </div>
    <div
      class="title title-2-default "
      id="two-title"
      role="button"
      tabindex="0"
    >
      Title 2
    </div>
    <div
      aria-hidden="true"
      class="panel collapsed panel-2-default "
      id="two-panel"
      style="height: 0px;"
    >
      Panel 2
    </div>
  </div>
</div>
`;
